merkle-tree/concurrent/tests/tests.rs

```diff
@@ -2204,3 +2220,698 @@ fn test_subtree_updates() {
+        .collect::<Vec<_>>();
+    assert!(changelog_entries.is_empty());
+
+    // Vector of leaves we append and update.
+    let mut leaves = CyclicBoundedVec::with_capacity(CHANGELOG);
+    // Changelog is always initialized with a changelog path consisting of zero
+    // bytes. For consistency, we need to assert the 1st zero byte as the first
+    // expected leaf in the changelog.
+    leaves.push(H::zero_bytes()[0]);
+
+    for _ in 0..1000 {
+        // Append random leaf.
+        let leaf: [u8; 32] = Fr::rand(&mut rng)
+            .into_bigint()
+            .to_bytes_be()
+            .try_into()
+            .unwrap();
+        merkle_tree.append(&leaf).unwrap();
+        reference_tree.append(&leaf).unwrap();
+        leaves.push(leaf);
+
+        let changelog_entries = merkle_tree
+            .changelog_entries(merkle_tree.changelog.first_index())
+            .collect::<Vec<_>>();
+        assert_eq!(changelog_entries.len(), leaves.len() - 1);
+
+        for (leaf, changelog_entry) in leaves.iter().skip(1).zip(changelog_entries) {
+            assert_eq!(&changelog_entry.path[0], leaf);
+        }
+
+        // Requesting changelog entries starting from the current `changelog_index()`
+        // should always return an empty iterator.
+        let changelog_entries = merkle_tree
+            .changelog_entries(merkle_tree.changelog_index())
+            .collect::<Vec<_>>();
+        assert!(changelog_entries.is_empty());
+
+        // Update random leaf.
+        let leaf_index = rng.gen_range(0..reference_tree.leaves().len());
+        let old_leaf = reference_tree.leaf(leaf_index);
+        let new_leaf: [u8; 32] = Fr::rand(&mut rng)
+            .into_bigint()
+            .to_bytes_be()
+            .try_into()
+            .unwrap();
+        let mut proof = reference_tree.get_proof_of_leaf(leaf_index, false).unwrap();
+        merkle_tree
+            .update(
+                merkle_tree.changelog_index(),
+                &old_leaf,
+                &new_leaf,
+                leaf_index,
+                &mut proof,
+            )
+            .unwrap();
+        reference_tree.update(&new_leaf, leaf_index).unwrap();
+        leaves.push(new_leaf);
+        println!("Updated leaf {leaf_index} {old_leaf:?} to {new_leaf:?}");
+
+        println!("\n\n\nLEAVES:");
+        for (i, leaf) in leaves.iter().enumerate() {
+            println!("{i}: {leaf:?}");
+        }
+        println!("\n\n\nCHANGELOG:");
+        for (i, changelog_entry) in merkle_tree.changelog.iter().enumerate() {
+            println!("{i}: {:?}", changelog_entry.path[0]);
+        }
+
+        let changelog_entries = merkle_tree
+            .changelog_entries(merkle_tree.changelog.first_index())
+            .collect::<Vec<_>>();
+        assert_eq!(changelog_entries.len(), leaves.len() - 1);
+
+        for (leaf, changelog_entry) in leaves.iter().skip(1).zip(changelog_entries) {
+            assert_eq!(&changelog_entry.path[0], leaf);
```

Comment:
can we assert the complete path here?

Reply:
can we also assert the other fields of the changelog object?

Reply:
Done